internal/application/services.go

```diff
@@ -15,6 +17,133 @@ import (
 
 type NameSet map[string]struct{}
 
+// BoundApps is an extension of BoundAppsNames after it, to retrieve a map of services to
+// the full data of the applications bound to them. It uses BoundAppsNames internally to
+// quickly determine the applications to fetch.
+
+func BoundApps(ctx context.Context, cluster *kubernetes.Cluster, namespace string) (map[string]models.AppList, error) {
+
+	result := map[string]models.AppList{}
+
+	bindings, err := BoundAppsNames(ctx, cluster, namespace)
+	if err != nil {
+		return result, err
+	}
+
+	// Internal map of fetched applications.
+	fetched := map[string]*models.App{}
+
+	for key, appNames := range bindings {
+		serviceName, namespace := DecodeServiceKey(key)
+		for _, appName := range appNames {
+			app, ok := fetched[appName]
+			if !ok {
+				meta := models.NewAppRef(appName, namespace)
+				app := meta.App()
+				err := fetch(ctx, cluster, app)
+				if err != nil {
+					// Ignoring the error. Assumption is that
+					// the app got deleted as this function is
+					// collecting its information.
+					break
+				}
+				fetched[appName] = app
+			}
+			result[serviceName] = append(result[serviceName], *app)
+		}
+	}
+
+	return result, nil
+}
+
+// BoundAppsNamesFor is a specialization of BoundAppsNames after it, to retrieve the names
+// of the applications bound to a single service, specified by name.
+
+func BoundAppsNamesFor(ctx context.Context, cluster *kubernetes.Cluster, namespace, serviceName string) ([]string, error) {
+	result := []string{}
+
+	// locate service bindings managed by epinio applications
+	selector := EpinioApplicationAreaLabel + "=service"
+	selector += ",app.kubernetes.io/component=application"
+	selector += ",app.kubernetes.io/managed-by=epinio"
+
+	appBindings, err := cluster.Kubectl.CoreV1().Secrets(namespace).List(ctx,
+		metav1.ListOptions{
+			LabelSelector: selector,
+		})
+	if err != nil {
+		return result, err
+	}
+
+	// Instead of building a full inverted map from service names to app names here we
+	// filter on the service name to generate just that slice of app names.
+
+	for _, binding := range appBindings.Items {
+		for boundServiceName := range binding.Data {
+			if boundServiceName == serviceName {
+				appName := binding.ObjectMeta.Labels["app.kubernetes.io/name"]
+				result = append(result, appName)
+				break
+			}
+		}
+	}
+
+	return result, nil
+}
+
+// BoundAppsNames returns a map from the service names of services in the specified namespace,
+// to the names of the applications they are bound to. For an empty namespace the map
+// contains the services in all namespaces, and the keys are a combination of namespace
```

Comment:
Update comment? (now it's always a combination of name and namespace)

Reply:
Right. Just a moment.

Reply:
commit 2903a9968e0a1a0ad8c2859cad9275f7ab96551b